packages/core/src/domain/conversion/bundle-modifications/modifications.ts

```diff
@@ -0,0 +1,106 @@
+import { Bundle, Resource } from "@medplum/fhirtypes";
+import { cloneDeep } from "lodash";
+import * as uuid from "uuid";
+import { isPatient } from "../../../external/fhir/shared";
+import { DOC_ID_EXTENSION_URL } from "../../../external/fhir/shared/extensions/doc-id-extension";
+
+export type FhirExtension = {
+  url: string;
+  valueString: string;
+};
+
+export type FhirConverterParams = {
+  patientId: string;
+  fileName: string;
+  unusedSegments: string | undefined;
+  invalidAccess: string | undefined;
+};
+
+/**
+ * Replaces the IDs (and all references to them) for all resources that have an extension for the source document.
+ * This is done so that identical resources originated from different sources would have different IDs.
+ */
+export function replaceIdsForResourcesWithDocExtension(
+  fhirBundle: Bundle<Resource>,
+  patientId: string
+): Bundle<Resource> {
+  const updatedBundle = cloneDeep(fhirBundle);
+  const stringsToReplace: { old: string; new: string }[] = [];
+  if (!updatedBundle.entry || updatedBundle.entry.length === 0) {
+    throw new Error(`Missing bundle entries`);
+  }
+  for (const bundleEntry of updatedBundle.entry) {
+    if (!bundleEntry.resource) throw new Error(`Missing resource`);
+    if (!bundleEntry.resource.id) throw new Error(`Missing resource id`);
+    if (bundleEntry.resource.id === patientId) continue;
+
+    const resource = bundleEntry.resource;
+    if ("extension" in resource) {
+      const docIdExtension = resource.extension?.find(ext => ext.url === DOC_ID_EXTENSION_URL);
+      const idToUse = bundleEntry.resource.id;
+      const newId = uuid.v4();
+      stringsToReplace.push({ old: idToUse, new: newId });
+      // replace meta's source and profile
+      bundleEntry.resource.meta = {
+        lastUpdated: bundleEntry.resource.meta?.lastUpdated ?? new Date().toISOString(),
+        source: docIdExtension?.valueString ?? "",
+      };
+    }
+  }
+  let updatedBundleStr = JSON.stringify(updatedBundle);
+  for (const stringToReplace of stringsToReplace) {
+    // doing this is apparently more efficient than just using replace
```

Comment:
That's cool. Let's then move this to an entry on the code standards with a link to a reference material, please?

Reply:
this comment is also extremely old. It was already there on [this PR](https://github.com/metriport/metriport/pull/1442)

Reply:
so if you don't mind, ill just remove it, so i dont end up spending time looking for articles that someone else used when writing this code

Reply:
I was curious and did some tests (available in [this gist](https://gist.github.com/leite08/9ebe4cadca90a04f6ca7870e005c3a4d)) and also found [this article](https://christiankohler.net/it-s-a-trap-the-biggest-pitfall-of-string-prototype-replace#:~:text=Global_Objects/String/replaceAll-,Performance,-.%20Both%20functions%20are) that mentions similar results.

`.replace()` with a regex is faster - but with the inline one, not creating a new `RegExp` instance on each call (that's bad for performance, unless we're reusing the same regex in all calls, which is not the case there).